### PR TITLE
[Small Feature] Point Set Clustering

### PR DESCRIPTION
**Pull request by sgiraudot**
## Rationale

Identifying clusters of locally connected points in a point set is a pretty standard procedure. Although the algorithm is pretty simple and straightforward, CGAL does not provide any API to such an algorithm, which seems like a missing feature.

(In my personal experience, I used several times this algorithm and kept copy-pasting the code, so I figured it belonged in the _Point Set Processing_ package.)

This small features adds a clustering algorithm to _Point Set Processing_.

## Summary of API changes

A new function `CGAL::cluster_point_set()` is provided, along with a new named parameter `adjacencies`.

## License and copyright ownership

GPL, GeometryFactory

## CHANGES.md

(todo)

### Version 1 (_outdated_)

* [~~__User manual__~~](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set/Point_set_processing_3/index.html#title17)
* [~~__Reference manual__~~](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga636fa37aa7ed4ae296028d4b67470a24)

### Version 2

* [__User manual__](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set_v2/Point_set_processing_3/index.html#title17)
* [__Reference manual__](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set_v2/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#gafee41d60b5a257ae034e9157d0af8e46)

## Status

- Developed and locally tested (GNU/Linux)
- [Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Point_Set_Clustering)  --  pre-approved Andreas Fabri 10:50, 8 April 2020 (CEST) 
- __Note:__ this features depends on https://github.com/CGAL/cgal/pull/4552, so it should be integrated afterwards (I'm not really worried, small features usually take longer to be integrated than simple PR)

### CODE REVIEW

**Comment by mglisse** [comment]
Precisely because clustering is such a well-known and wide topic, it should be clear what algorithm you are using. Are you just returning the connected components of a union of balls? The generic name "clustering" may give users a lot of expectations.

**Comment by sgiraudot** [comment]
The documentation says:

> Identifies simply connected clusters on a nearest neighbors graph. 

How can I make it clearer?

**Comment by mglisse** [comment]
simply connected clusters -> connected components

nearest neighbor graph: that has a precise definition. Neighborhood graph is more vague. I've seen the graph with all pairs at distance less than r called the r-Rips graph, proximity graph, maybe also some name with "ball" in it, not sure about it, it probably doesn't have a canonical name. Anyway, a sentence explaining what it does cannot hurt.

**Comment by afabri** [comment]
I suggest to add an image for a 2D point set (then it is planar which is better for html).  
I think your "neighbor" function must be symmetric as you strive for equivalence classes.  It does not seem to be well defined for functions like  "a link to the closest point"  which are not symmetric.

**Comment by maxGimeno** [comment]
Errors on travis [here](https://travis-ci.com/github/CGAL/cgal/jobs/296517761#L3502) and [here](https://travis-ci.com/github/CGAL/cgal/jobs/296517732#L7434)

**Comment by sgiraudot** [comment]
I uploaded a new version of the doc following the different reviews I got. Following a discussion with @afabri, I removed the parameter `K`, as using a number of nearest neighbors to create the nearest neighbor graph creates a directed graph that is not well suited for this algorithm. Now only a neighbor radius can be used, which makes the graph undirected and the algorithm better defined. I also added a figure.

**Comment by MaelRL** [comment]
@afabri @sgiraudot bump

**Comment by maxGimeno** [comment]
Warning : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-127/Polyhedron_Demo/TestReport_lrineau_Ubuntu-Latest-CXX11.gz

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-130.shtml

**Comment by sgiraudot** [comment]
Is this something I can do? Will this be fixed if I add the Property_map package to the dependency files of packages listed in:
```
Differences in Voronoi_diagram_2: Property_map are new and not committed.
 Differences in Triangulation: Property_map are new and not committed.
 Differences in TDS_3: Property_map are new and not committed.
 Differences in TDS_2: Property_map are new and not committed.
 Differences in STL_Extension: Property_map are new and not committed.
 Differences in QP_solver: Property_map are new and not committed.
 Differences in Polynomial: Property_map are new and not committed.
 Differences in Point_set_3: Circulator are new and not committed.
 Differences in Mesh_2: Property_map are new and not committed.
 Differences in Generalized_map: Property_map are new and not committed.
 Differences in Convex_hull_d: Property_map are new and not committed.
 Differences in Cone_spanners_2: Property_map are new and not committed.
 Differences in Bounding_volumes: Property_map are new and not committed.
 Differences in Boolean_set_operations_2: Property_map are new and not committed.
 Differences in Algebraic_kernel_d: Property_map are new and not committed.
```

**Comment by maxGimeno** [comment]
@sgiraudot yes, exactly

**Comment by sgiraudot** [comment]
I ended up moving the new functions to `Property_map` which seem cleaner and shouldn't change the dependencies.

**Comment by maxGimeno** [comment]
Circulator is a new dep of Point_set_3, apparently.

**Comment by maxGimeno** [comment]
warnings in the demo : 
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-137/Polyhedron_Demo/TestReport_lrineau_Ubuntu-latest-GCC6-Release.gz

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-139.shtml.

**Comment by sloriot** [comment]
@sgiraudot could you please update `changes.md`?

File: Property_map/include/CGAL/property_map.h [hunk 1]
```diff
@@ -576,6 +578,38 @@ make_cartesian_converter_property_map(Vpm vpm)
...
+  return CGAL::make_range
+    (make_transform_iterator_from_property_map (range.begin(), pmap),
+     make_transform_iterator_from_property_map (range.end(), pmap));
+}
+
+// Syntaxic sugar for make_range+transform_iterator+pmap_to_unary_function
+template <typename Range, typename Pmap>
+CGAL::Iterator_range<typename boost::transform_iterator<CGAL::Property_map_to_unary_function<Pmap>,
+                                                        typename Range::iterator> >
+make_transform_range_from_property_map (Range& range, Pmap pmap)
```
Comment (sloriot) [inline Property_map/include/CGAL/property_map.h:605]: I would rename that to `transform_range_[with/using]_property_map()`

File: Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      neighbor_radius->setValue(-1);
+    }
+
+    // Computes average spacing
+    Clustering_functor functor (points, neighbor_radius->value(), cluster_map);
+    run_with_qprogressdialog (functor, "Clustering...", mw);
+
+    std::size_t nb_clusters = *functor.result;
+
+    Scene_group_item* group;
```
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:145]: warning: may be used unitialized

File: Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    if (gen_sub->isChecked())
+    {
+      group = new Scene_group_item(QString("%1 (clusters)").arg(item->name()));
+      scene->addItem(group);
+      new_items.reserve (nb_clusters);
+      for (std::size_t i = 0; i < nb_clusters; ++ i)
+      {
+        Scene_points_with_normal_item* new_item = new Scene_points_with_normal_item;
+        new_item->point_set()->copy_properties (*points);
+        CGAL::Random rand(i);
```
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:157]: conversion from 'size_t' to 'unsigned int

File: Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+      colored->point_set()->check_colors();
+
+      colored->point_set()->reserve (points->size());
+
+      for (Point_set::Index idx : *points)
+      {
+        Point_set::Index iidx = *(colored->point_set()->insert (points->point(idx)));
+        if (cluster_size[cluster_map[idx]] >= std::size_t(min_nb->value()))
+        {
+          CGAL::Random rand(cluster_map[idx] + 1);
```
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:196]: conversion from 'size_t' to 'unsigned int

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-152.shtml

**Pull request merged**